snappy/snap_remote_repo.go

```diff
@@ -383,3 +386,61 @@ func (s *SnapUbuntuStoreRepository) Updates() (parts []Part, err error) {
 func (s *SnapUbuntuStoreRepository) Installed() (parts []Part, err error) {
 	return nil, err
 }
+
+// Download downloads the given snap and returns the filename of the local
+// version
+func (s *SnapUbuntuStoreRepository) Download(remoteSnap *RemoteSnapPart, pbar progress.Meter) (string, error) {
+	w, err := ioutil.TempFile("", remoteSnap.pkg.Name)
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		if err != nil {
+			os.Remove(w.Name())
+		}
+	}()
+	defer w.Close()
+
+	// try anonymous download first and fallback to authenticated
+	url := remoteSnap.pkg.AnonDownloadURL
+	if url == "" {
+		url = remoteSnap.pkg.DownloadURL
+	}
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return "", err
+	}
+	setUbuntuStoreHeaders(req)
+
+	if err := download(remoteSnap.Name(), w, req, pbar); err != nil {
+		return "", err
+	}
+
+	return w.Name(), w.Sync()
```

Comment:
I'm not 100% sure that syncing covers has all the important low-level semantics of closing the file. For the typical case, probably, but it sounds so easy to have it clearly right.

There's also another issue about the w.Name() return here. The defer removes the file, but the call site still gets the name pointing to a non-existent file. Instead, this function should be declared with named results (e.g. `(path string, err error)`), and the path reset after the file is removed on errors.